src/controller/proxy/manifestcache.go

```diff
@@ -0,0 +1,238 @@
+//  Copyright Project Harbor Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/manifest/manifestlist"
+	"github.com/docker/distribution/manifest/schema2"
+	"github.com/goharbor/harbor/src/lib"
+	libCache "github.com/goharbor/harbor/src/lib/cache"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/opencontainers/go-digest"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"strings"
+	"time"
+)
+
+const defaultHandler = "default_handler"
+
+// NewCacheHandlerRegistry ...
+func NewCacheHandlerRegistry(local localInterface) map[string]ManifestCacheHandler {
+	registry := map[string]ManifestCacheHandler{
+		manifestlist.MediaTypeManifestList: &ManifestListCache{local: local, cache: libCache.Default(), listProcessor: ListProcessor{local, libCache.Default()}},
+		v1.MediaTypeImageIndex:             &ImageIndexCache{local, ListProcessor{local, libCache.Default()}},
+		schema2.MediaTypeManifest:          &ManifestCache{local},
+		"default":                          &ManifestCache{local},
+	}
+	return registry
+}
+
+// ManifestCacheHandler define how to cache manifest content
+type ManifestCacheHandler interface {
+	// CacheContent - cache the content of the manifest
+	CacheContent(ctx context.Context, remoteRepo string, man distribution.Manifest, art lib.ArtifactInfo, r RemoteInterface)
+	// Push - push the manifest to local registry
+	Push(ctx context.Context, repo, reference string, man distribution.Manifest) error
+}
+
+// ManifestListCache handle Manifest list type
+type ManifestListCache struct {
```

Comment:
Why there two separated implementation for manifest list and index? They are almost the same thing except the media type

Reply:
The implementation of CacheContent are different.